### PR TITLE
Avoid initialization of relcache during recovery

### PR DESCRIPTION
**Pull request by xinzweb**
This patch set defers `relcache` initialization in crash recovery as well as filerep processes until after xlog replay. If `relcache` is initialized before xlog replay, this leads to inconsistencies such as results of a committed transaction lost after recovery.

A error message put in place to expose this issue. This will prevent any future regression.

### CODE REVIEW

File: src/backend/utils/cache/relcache.c [hunk 1]
```diff
@@ -3370,6 +3370,13 @@ RelationCacheInitializePhase3(void)
 	bool		needNewCacheFile = !criticalSharedRelcachesBuilt;
 
 	/*
+	 * Relation cache initialization or any sort of heap access is
+	 * dangerous before recovery is finished.
+	 */
+	if (!IsBootstrapProcessingMode() && RecoveryInProgress())
+		elog(ERROR, "relation cache initialization during recovery or non-boostrap processes.");
```
Comment (danielgustafsson) [inline src/backend/utils/cache/relcache.c]: Typo, s/boostrap/bootstrap/

Comment (xinzweb) [inline src/backend/utils/cache/relcache.c]: thanks, fixed.

File: src/backend/access/transam/xlog.c [hunk 2]
```diff
@@ -7725,6 +7725,35 @@ void
...
+	else
+	{
+		MyDatabaseId = XLogCtl->currentDatabaseToVerify;
+		MyDatabaseTableSpace = XLogCtl->tablespaceOfCurrentDatabaseToVerify;
+	}
+
+	/*
+	 * Now we can mark our PGPROC entry with the database ID */
+	/* (We assume this is an atomic store so no lock is needed)
+	 */
```
Comment (xinzweb) [inline src/backend/access/transam/xlog.c]: thanks, fixed.

**Comment by ashwinstar** [comment]
Merged.

**Pull request closed without merging**